src/bootupd.rs

```diff
@@ -489,6 +496,173 @@ pub(crate) fn client_run_validate() -> Result<()> {
+    ensure_writable_boot()?;
+
+    let grub_config_dir = PathBuf::from("/boot/grub2");
+    let dirfd = openat::Dir::open(&grub_config_dir).context("Opening /boot/grub2")?;
+
+    // We mark the bootloader as BLS capable to disable the ostree-grub2 logic.
+    // We can do that as we know that we are run after the bootloader has been
+    // updated and all recent GRUB2 versions support reading BLS configs.
+    // Ignore errors as this is not critical. This is a safety net if a user
+    // manually overwrites the (soon) static GRUB config by calling `grub2-mkconfig`.
+    // We need this until we can rely on ostree-grub2 being removed from the image.
+    println!("Marking bootloader as BLS capable...");
+    _ = File::create("/boot/grub2/.grub2-blscfg-supported");
+
+    // Migrate /boot/grub2/grub.cfg to a static GRUB config if it is a symlink
+    let grub_config_filename = PathBuf::from("/boot/grub2/grub.cfg");
+    match dirfd.read_link("grub.cfg") {
+        Err(_) => {
+            println!(
+                "'{}' is not a symlink, nothing to migrate",
+                grub_config_filename.display()
+            );
+        }
+        Ok(path) => {
+            println!("Migrating to a static GRUB config...");
+
+            // Resolve symlink location
+            let mut current_config = grub_config_dir.clone();
+            current_config.push(path);
+
+            // Backup the current GRUB config which is hopefully working right now
+            let backup_config = PathBuf::from("/boot/grub2/grub.cfg.backup");
+            println!(
+                "Creating a backup of the current GRUB config '{}' in '{}'...",
+                current_config.display(),
+                backup_config.display()
+            );
+            fs::copy(&current_config, &backup_config).context("Failed to backup GRUB config")?;
+
+            // Read the current config, strip the ostree generated GRUB entries and
+            // write the result to a temporary file
+            println!("Stripping ostree generated entries from GRUB config...");
+            let current_config_file = File::open(current_config).context("Could not open current GRUB config")?;
+            let stripped_config = String::from("grub.cfg.stripped");
+            // mode = -rw-r--r-- (644)
+            let mut writer = BufWriter::new(dirfd.write_file(&stripped_config,
+                (S_IRUSR | S_IWUSR | S_IRGRP | S_IROTH) as mode_t)
+                .context("Failed to open temporary GRUB config")?
+            );
+            let mut skip = false;
+            for line in BufReader::new(current_config_file).lines().map_while(Result::ok) {
+                if line == "### END /etc/grub.d/15_ostree ###" {
+                    skip = false;
+                }
+                if skip {
+                    continue;
+                }
+                if line == "### BEGIN /etc/grub.d/15_ostree ###" {
+                    skip = true;
+                }
```

Comment:
No need to keep "### BEGIN /etc/grub.d/15_ostree ###" in the new file
```suggestion
                if line == "### BEGIN /etc/grub.d/15_ostree ###" {
                    skip = true;
                }
                if line == "### END /etc/grub.d/15_ostree ###" {
                    skip = false;
                }
                if skip {
                    continue;
                }
```

Reply:
This will leave the END line only, which is inconsistent. The logic is simpler to understand if I keep both lines.